### PR TITLE
search: mint search routine type

### PR DESCRIPTION
**Pull request by rvantonder**
This mints a `Routine` type to hold the 3-tuple values previously used to evaluate a toplevel search:

```go
type Routine struct {
	Jobs        []Job
	RepoOptions search.RepoOptions
	Timeout     time.Duration
}
```

After thinking about a name for this type and settling on `Routine`, and propagating it, I think it works well. See docstring for more details.

This is semantics-preserving. Comments and functions are updated to reflect this type. This structure is generally more comprehensible, but there's another motivation. Optional reading below if you want to understand why.

---

The main motivation is that I want to move from our current `evaluate*(...)` functions that operate on `query.Basic`, to instead evaluate on `run.Routine`. That is, a search query plan gets converted to a tree of `run.Routine`. This conversion phase is needed if we want to do optimizations (like Zoekt `and`) cleanly. Currently there's an issue with the order in which search jobs/routines are evaluated. If we implemented the Zoekt optimization now, we'd effectively have to go and find search routines containing zoekt jobs (where the work has already been split up) and replace it with the Zoekt-optimized job, OR, we'd have to add the Zoeekt-optimized job to a routine, and then pass some conditional so that subsequent `and` logic does not also add unoptimized Zoekt jobs. If this doesn't quite make sense, don't worry about it. The point is that the eval functions operate on a `query.Basic` type and impose a certain ordering of job/routine creation that makes it difficult to modify the output of that creation for optimized backend queries. The solution is to not operate on `query.Basic` during eval, but instead convert it before eval to a tree of routines. The intent has always been to not evaluate on `query.Basic` and I made a comment about this ages ago--we want to evaluate on our internal type (best represented by `Routine` now). 

There is no reason we can't turn a query into a tree of `Routines`, and, in fact, our `type Plan []query.Basic` is going to eventually be `type Plan <tree of routines>`. Actually that will probably happen pretty soon, because that representation is conducive to the optimization phase. Keegan has brought up the idea of an "IR" for our search code, and `type Plan <tree of routines>` is currently our best representation of that "IR". Notably that representation of a backend plan should not (and should never have had to) care directly about anything in the query string.

### CODE REVIEW

File: internal/search/run/run.go [hunk 1]
```diff
@@ -43,6 +42,20 @@ type Job interface {
...
+// or return different result types (text vs. symbols). The relation with
+// SearchInputs and Routine is that SearchInputs are static values, parsed and
+// validated, to produce one or more Routines. Routines express the complete
+// information to execute the runtime semantics for particular search
+// operations.
+type Routine struct {
+	Jobs        []Job
+	RepoOptions search.RepoOptions
+	Timeout     time.Duration
+}
```
Comment (camdencheek) [inline internal/search/run/run.go:57]: What stops us from making `Routine` implement the `Job` interface? Is it just that `searchrepos.Pager` is a runtime value that can't be created "statically"? Just thinking that a `Job` that runs child `Job`s in parallel would be elegant here.

**Review by camdencheek — APPROVED** [review]
I really love where this is going. I'm extremely excited for the "static tree of routines" IR. That would make all this code monitor stuff I'm doing right now _soooo_ much easier (can mutate the tree ahead of time rather than changing behavior at runtime).

File: internal/search/run/run.go (hunk 1 above)
Comment (rvantonder) [reply internal/search/run/run.go:57]: Hey good question! Didn't think about this before. Thinking through it right now, practically I don't think there's something stopping us. This _would_ let us change `doResults` to just be a `job.Run` which is very elegant. Let me try create another PR so we can compare. The thing I hesitate on is that it risks conflating things. Looking at the Job interface, I think the following would be cruft (which are "orange" flags suggesting the concerns are not the same):

- `Required()` -> noop, we will never look at this value because our code structure implies a routine (or if we represented it with job, this job) is always executed.

- `searchrepos.Pager` -> would be ignored, we'd create this inside the job. Not a disaster since we ignore this for some search jobs. To your question, maybe it's possible to create this before hand, but I'm not sure and it makes my head hurt. But like I say practically speaking we'd just ignore this value.

Basically, I'm unsure if `Job` as reflected in our code structure is generic enough to also represent `Routine`.

Comment (camdencheek) [reply internal/search/run/run.go:57]: > `Required()` -> noop

I still think we can get rid of the whole "optional" system with streaming now, but I've not actually done the work to verify that.

Another option, we could have a job `TieredJob` that handles the current concerns of `Required()`. It could looks something like this:
```go
type TieredJob struct {
    required []Job
    optional []Job
}
```

> `searchrepos.Pager`

I don't love ignoring it either. Maybe we change the job interface to not require `searchrepos.Pager`, then do we can have something like `ParallelJob`:

```go
type JobWithoutPager interface {
    WithPager(searchrepos.Pager) Job    
}

type ParallelJob struct {
    jobsWithoutPagers []func(searchrepos.Pager) 
}

func (p ParallelJob) Run(ctx context.Context, s stream) error {
    pager := createPager(ctx) // this is a runtime thing
    for _, pagerlessJob := range p.jobsWithoutPagers {
      go func() {
          return pagerlessJob.WithPager(pager).Run(ctx, s)
       }()
    }
    ...
}
```

Comment (rvantonder) [reply internal/search/run/run.go:57]: > This would let us change `doResults` to just be a `job.Run` which is very elegant.

Ah okay, so there's some real trouble here--to do this, this toplevel job/routine would have access to various things in `searchResolver`. I think we lose the elegance of this solution if we make `searchResolver` part of this job's args 😬 . To do it cleanly, we need to factor out the dependencies on `searchResolver` (like `r.Db` and various functions). Which is feasible and desirable, but would take some effort and my focus is to get us to separating the `query` <-> `routine+jobs` concern rather than merging `routine+jobs` representation. I think we should come back to this idea (and continue trimming search resolver deps).

Comment (camdencheek) [reply internal/search/run/run.go:57]: Ah, yes. That makes a lot of sense. Am in full support of _not_ making the resolver part of the job's args

Comment (rvantonder) [reply internal/search/run/run.go:57]: Tis a pity, I think it pans out well otherwise. Tracking in https://github.com/sourcegraph/sourcegraph/issues/30058 because I'm going to have trouble finding this thread in the future :3
